In [44]:
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf
import time

In [45]:
from keras import models 
from keras import layers
from keras.utils import to_categorical

In [46]:
from keras.datasets import mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()

In [47]:
# Data normalization
X_train = X_train.reshape((60000, 28*28))
X_train = X_train/255

X_test = X_test.reshape((10000, 28*28))
X_test = X_test/255

In [48]:
# y_train, y_test data preprocessing
y_train=to_categorical(y_train)
y_test=to_categorical(y_test)

In [49]:
#modeling

1.Softmax regression

In [50]:
start_soft = time.time()
net_soft = models.Sequential()
net_soft.add(layers.Dense(10, activation='softmax', input_shape=(28*28,)))
net_soft.compile(optimizer='adam',loss = 'categorical_crossentropy',metrics=['accuracy'])

In [51]:
net_soft.fit(X_train, y_train, epochs=20)

Epoch 1/20
60000/60000 [==============================] - 3s 53us/step - loss: 0.4694 - accuracy: 0.8772
Epoch 2/20
60000/60000 [==============================] - 3s 43us/step - loss: 0.3043 - accuracy: 0.9151
Epoch 3/20
60000/60000 [==============================] - 2s 37us/step - loss: 0.2836 - accuracy: 0.9209
Epoch 4/20
60000/60000 [==============================] - 2s 38us/step - loss: 0.2730 - accuracy: 0.9237
Epoch 5/20
60000/60000 [==============================] - 2s 41us/step - loss: 0.2668 - accuracy: 0.9260
Epoch 6/20
60000/60000 [==============================] - 2s 38us/step - loss: 0.2621 - accuracy: 0.9274
Epoch 7/20
60000/60000 [==============================] - 2s 38us/step - loss: 0.2583 - accuracy: 0.9284
Epoch 8/20
60000/60000 [==============================] - 2s 40us/step - loss: 0.2555 - accuracy: 0.9293
Epoch 9/20
60000/60000 [==============================] - 2s 39us/step - loss: 0.2529 - accuracy: 0.9298
Epoch 10/20
60000/60000 [==============================

In [52]:
soft_loss, soft_acc = net_soft.evaluate(X_test,y_test)

10000/10000 [==============================] - 0s 34us/step


In [53]:
# softmax 방식으로 얻은 accuracy
print(soft_acc)
print('softmax time:', time.time()-start_soft)

0.9273999929428101
softmax time: 48.67577815055847


2. Support Vector Machine

In [65]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler

(X_svm_train, y_svm_train), (X_svm_test, y_svm_test) = mnist.load_data()

scaler = StandardScaler()
# Data normalization
X_svm_train = X_svm_train.reshape((60000, 28*28))
X_svm_train = X_svm_train/255

X_svm_test = X_svm_test.reshape((10000, 28*28))
X_svm_test = X_svm_test/255

In [66]:
# Data scaling 
X_svm_train_scaled = scaler.fit_transform(X_train.astype(np.float32))
X_svm_test_scaled = scaler.fit_transform(X_test.astype(np.float32))

In [67]:
# modeling
start_svm = time.time()
svc=SVC(kernel='linear')
svc.fit(X_svm_train,y_svm_train)

SVC(kernel='linear')

In [68]:
# svm 방식으로 얻은 accuracy
y_svm_pred = svc.predict(X_svm_test)
print(accuracy_score(y_svm_pred, y_svm_test))
print('svm time:', time.time()-start_svm)

0.9404
svm time: 591.6522121429443


3. RandomForest

In [69]:
#modeling
from sklearn.ensemble import RandomForestClassifier
clf  =  RandomForestClassifier(n_estimators = 200, n_jobs = -1)
start_rf = time.time()

In [70]:
clf.fit(X_train,y_train)

RandomForestClassifier(n_estimators=200, n_jobs=-1)

In [71]:
# Randomforest 방식으로 얻은 accuracy
y_clf_pred = clf.predict(X_test)
print(accuracy_score(y_clf_pred, y_test))
print('rf time:', time.time()-start_rf)

0.9047
rf time: 83.41883873939514


4. 결과 분석 및 고찰 
softmax regression으로 얻은 정확도는 0.9265
svm으로 얻은 정확도는 0.9404
RandomForest로 얻은 정확도는 0.9043 이다. 
각각의 방법이 완전히 동일한 조건으로 둔 것이 아니라 이들을 단순히 정확도로 무엇이 더 우세하다고 말하긴 힘들다.
하지만 각각의 방법들의 계산 속도를 보면 softmax 방식은 48초, svm은 대략 10분정도 소요했고, radomforest는 1분 40초가 소요되었다.
아무리 높은 정확도를 가진 방법이라 해도 한번 시행에 너무 많은 시간이 걸리는 svm은 이 경우에는 비합리적이라 판단되어진다.
따라서 이 경우 softmax 방식을 사용하는 것이 다른 두가지 방법보다는 좀더 합리적인 판단이라고 생각 된다. 